src/memory.jl

```diff
@@ -11,6 +11,19 @@ using Base.Threads: SpinLock
 # each allocator needs to lock its own resources separately too.
 const memory_lock = SpinLock()
 
+# the above spinlocks are taken around code that might gc, which might cause a deadlock
+# if we try to acquire from the finalizer too. avoid that by temporarily disabling the GC.
+macro without_gc(ex)
+  quote
+    gc = GC.enable(false)
```

Comment:
Julia base uses `jl_gc_enable_finalizers(ptls, 0)`.

```
ptls = Core.getptls()
ccall(:jl_gc_enable_finalizers, Cvoid, (Ptr{Cvoid}, Cint), ptls, 0)
```

Should be enough to disable finalizers to run. There is a question if the code needs to be sigatomic.
Looking at the code in locks.h I am a bit confused. It is unclear to me what happens when two threads try to acquire two different mutexes and they interleave the calls to `jl_gc_enable_finalizers`, similarly here we can't really guard against some-one enabling GC while we are in the critical section.